state/postgresql/migrations.go

```diff
@@ -0,0 +1,227 @@
+/*
+Copyright 2021 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package postgresql
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/dapr/kit/logger"
+)
+
+// Performs migrations for the database schema
+type migrations struct {
+	Logger            logger.Logger
+	Conn              *sql.DB
+	StateTableName    string
+	MetadataTableName string
+}
+
+// Perform the required migrations
+func (m *migrations) Perform(ctx context.Context) error {
+	// Use an advisory lock (with an arbitrary number) to ensure that no one else is performing migrations at the same time
+	// This is the only way to also ensure we are not running multiple "CREATE TABLE IF NOT EXISTS" at the exact same time
+	// See: https://www.postgresql.org/message-id/CA+TgmoZAdYVtwBfp1FL2sMZbiHCWT4UPrzRLNnX1Nb30Ku3-gg@mail.gmail.com
+	const lockID = 42
+
+	// Long timeout here as this query may block
+	queryCtx, cancel := context.WithTimeout(ctx, time.Minute)
+	_, err := m.Conn.ExecContext(queryCtx, "SELECT pg_advisory_lock($1)", lockID)
```

Comment:
Just thinking aloud: Should we use `pg_try_advisory_lock` instead of `pg_advisory_lock` and return right away if lock already acquired by other session, and let migration be re-triggered later (need to give this part a thought, how re-trigger).

Reply:
I like that this blocks so we don't have to be concerned with retrying after a delay. Solves all the concurrency issues.